### PR TITLE
Write stages

### PR DESCRIPTION
**Pull request by CalumMcCall**
## What's new in this PR?

Write timestamps and stats for each stage

### CODE REVIEW

File: zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala [hunk 1]
```diff
@@ -17,46 +17,52 @@
...
 
-  override def markFCMNotificationsFetched(notifications: Seq[PushNotificationEncoded]): Future[Unit] = async {
-    val nots = notifications.map(_.id).toSet
-    val now = Instant.now
-    val fcmIds = await(filterFCMNotifications(nots))
-    val fut = fcmIds.map(id => storeNotificationState(id, FCMNotification.Fetched, now))
-    await(Future.sequence(fut))
+  override def markNotificationsWithState(ids: Set[Uid], stage: String): Future[Unit] =  async {
+    stage match {
+      case Pushed => await { Future.sequence(
```
Comment (yeryomenkom) [inline zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala]: Use `Future.traverse` in such case

File: zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala [hunk 2]
```diff
@@ -17,46 +17,52 @@
...
+        ids.toSeq.map { fcmTimestamps.storeNotificationState(_, stage, Instant.now) })}
+      case _ =>
+          val fcmIds = await(filterFCMNotifications(ids))
```
Comment (yeryomenkom) [inline zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala]: Use curly braces for `await` invocations.

File: zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala [hunk 3]
```diff
@@ -17,46 +17,52 @@
...
-  override def storeNotificationState(id: Uid, stage: String, timestamp: Instant): Future[Unit] =
-    async {
-      await { fcmTimestamps.storeNotificationState(id, stage, timestamp) }
-      val prevStage = await { fcmTimestamps.getPreviousStageTime(id, stage) }
-      if (prevStage.isDefined)
-        await { fcmStats.insertOrUpdate(getStageStats(stage, timestamp, prevStage.get)) }
-      if (stage == FCMNotification.FinishedPipeline) await { fcmTimestamps.deleteAllWithId(id) }
-    }
+  private def calcAndStore(ids: Set[Uid], stage: String): Future[Unit] = Future.sequence(ids.map { id =>
+    fcmTimestamps.getPreviousStageTime(id, stage).flatMap {
```
Comment (yeryomenkom) [inline zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala]: `Future.traverse`

File: zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala (hunk 1 above)
Comment (CalumMcCall) [reply zmessaging/src/main/scala/com/waz/service/FCMNotificationStatsService.scala]: Ah, I didn't realize `traverse` is simply a parallel version of `sequence`, thanks for the hint!

**Pull request merged**